Mathlib/Analysis/Convex/Radon.lean

```diff
@@ -0,0 +1,232 @@
+
+  have h5_I : ∑ i in I, w' i = 1 := by
+    let w'' : ι → 𝕜 := fun i => (w i) / weights_sum_I
+    have h5_aux : ∑ i in I, w' i = ∑ i in I, w'' i := by
+      apply Finset.sum_congr; rfl
+      intros i hi
+      rw [Finset.mem_filter] at hi
+      rcases hi with ⟨_, hh⟩
+      simp
+      contrapose; intros
+      exact Iff.mpr not_lt hh
+    rw [h5_aux]
+    simp
+    rw [←Finset.sum_div]
+    apply div_self
+    exact ne_of_gt h_I_pos
+
+  have h5_J : ∑ i in J, w' i = 1 := by
+    let w'' : ι → 𝕜 := fun i => (w i) / weights_sum_J
+    have h5_aux1 : weights_sum_I = -weights_sum_J := by
+      exact Iff.mp add_eq_zero_iff_eq_neg h3
+    have h5_aux2 : ∑ i in J, w' i = ∑ i in J, w'' i := by
+      apply Finset.sum_congr; rfl
+      intros i hi
+      rw [Finset.mem_filter] at hi
+      rcases hi with ⟨_, h_neg⟩
+      simp
+      split_ifs
+      exfalso; linarith
+      simp_rw [h5_aux1, div_neg]
+      ring
+    rw [h5_aux2]; simp
+    rw [←Finset.sum_div]
+    apply div_self
+    exact ne_of_lt h_J_neg
+
+  have h6 : ∀ v, w' v ≥ 0 := by
+    intro; simp; split_ifs
+    repeat {apply div_nonneg; linarith; linarith}
+
+  let p' : E := (Finset.sum J fun i => (w' i) • (f i))
+  have h7 : p = p' := by
+    suffices h7_suf1 : p -ᵥ p' = 0 by
+      exact eq_of_vsub_eq_zero h7_suf1
+
+    rw [Finset.weightedVSub_eq_weightedVSubOfPoint_of_sum_eq_zero s1 w f h_wsum 0]
+      at h_vsum
+    simp at h_vsum; simp
+
+    let g : ι → E := fun i =>  weights_sum_I⁻¹ • ((w i) • (f i))
+    have h7_aux2 : ∑ i in I, (w' i) • (f i) = ∑ i in I, g i := by
+      apply Finset.sum_congr; rfl
+      intros i hi
+      rw [Finset.mem_filter] at hi
+      rcases hi
+      simp
+      ring_nf; split_ifs
+      rw [mul_comm, mul_smul]
+      exfalso; linarith
+
+    have h7_aux3 : ∑ i in J, (w' i) • (f i) = -∑ i in J, g i := by
+      rw [←Finset.sum_neg_distrib]
+      apply Finset.sum_congr; rfl
+      intros i hi
+      rw [Finset.mem_filter] at hi
+      rcases hi
+      simp
+      ring_nf; split_ifs
+      exfalso; linarith
+      rw [mul_comm, ←neg_smul, ←neg_mul, mul_smul]
+
+    rw [h7_aux2, h7_aux3]
+    simp
+    have h7_aux4 := Finset.sum_filter_add_sum_filter_not s1 (fun i => 0 ≤ w i) g
+    simp at h7_aux4; rw [h7_aux4]
+    rw [←Finset.smul_sum, h_vsum, smul_zero]
+
+  use p
+  apply Set.mem_inter
+
+  apply Convex.sum_mem
+  exact convex_convexHull 𝕜 _
+  intros i _; exact h6 i
+  exact h5_I
+  intros i hi
+  have h8 : f i ∈ (f '' I) := by
+    exact Set.mem_image_of_mem f hi
+  apply Set.mem_of_mem_of_subset h8
+  apply subset_convexHull
+
+  rw [h7]
+  apply Convex.sum_mem
+  exact convex_convexHull 𝕜 _
+  intros i _; exact h6 i
+  exact h5_J
+  intros i hi
+  have h9 : f i ∈ (f '' Iᶜ) := by
+    apply Set.mem_image_of_mem
+    apply Set.mem_compl
+    intro h9_neg
+    have := (Iff.mp Finset.mem_filter hi).right
+    have := (Iff.mp Finset.mem_filter h9_neg).right
+    linarith
+  apply Set.mem_of_mem_of_subset h9
+  apply subset_convexHull
```

Comment:
Same comments as above. Note that in the second `apply`, the last `_` is the only one Lean can't guess, so it becomes the new goal.
```suggestion
  · rw [h7]
    apply (convex_convexHull 𝕜 _).sum_mem (fun i _ ↦ h6 i) h5_J
    intro i hi
    apply subset_convexHull _ _ (Set.mem_image_of_mem _ _)
    intro h9_neg
    exact not_lt_of_le (mem_filter.mp h9_neg).2 (mem_filter.mp hi).2
```